File: MarkdownEditor/MarkdownEditor/Sources/EditorViewController.m
//
//  EditorViewController.m
//  MarkdownEditor
//
//  Created by Iwaki Satoshi on 2018/02/27.
//  Copyright © 2018 Satoshi Iwaki and 2026 Vincent Janelle. All rights reserved.
//

#import "EditorViewController.h"
#import "ConverterManager.h"
#import "TextConverter.h"
#import "Editor/EditorDialogPresenter.h"
#import "Editor/EditorFileWatcher.h"
#import "Editor/EditorDocumentStore.h"
#import "Editor/EditorTextFormatter.h"

@interface EditorViewController () <NSTextViewDelegate> {
    NSString *_filePath;
    BOOL _dirty;
}

@property (unsafe_unretained) IBOutlet NSTextView *textView;
@property (unsafe_unretained) IBOutlet NSPopUpButton *converterPopup;
@property (nonatomic, strong) EditorDialogPresenter *dialogPresenter;
@property (nonatomic, strong) EditorFileWatcher *fileWatcher;

@end

@implementation EditorViewController

static const NSTimeInterval EditorAutosaveDelay = 0.5;
static const CGFloat EditorFormattingButtonWidth = 32.0;
static const CGFloat EditorFormattingButtonHeight = 25.0;
static const CGFloat EditorFormattingButtonGap = 8.0;
static const CGFloat EditorFormattingToolbarPadding = 6.0;
static const CGFloat EditorConverterPopupWidth = 170.0;
static const CGFloat EditorConverterPopupHeight = 25.0;

- (EditorDialogPresenter *)dialogPresenter {
    if (!_dialogPresenter) {
        _dialogPresenter = [[EditorDialogPresenter alloc] init];
    }
    return _dialogPresenter;
}

- (EditorFileWatcher *)fileWatcher {
    if (!_fileWatcher) {
        _fileWatcher = [[EditorFileWatcher alloc] init];
    }
    return _fileWatcher;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do view setup here.
    
    self.textView.accessibilityIdentifier = @"EditorTextView";
    self.converterPopup.accessibilityIdentifier = @"ConverterPopup";
    self.textView.string = [self loadSample];
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
}

- (void)viewDidLayout {
    [super viewDidLayout];
    [self layoutFormattingToolbar];
}

- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];
    
    // Update the view, if already loaded.
}

- (void)dealloc {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(autosaveIfNeeded) object:nil];
    [self.fileWatcher stopWatching];
}

#pragma mark - NSTextDelegate

- (void)textDidEndEditing:(NSNotification *)notification {
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
    _dirty = YES;
    [self scheduleAutosave];
}

- (void)textDidChange:(NSNotification *)notification {
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
    _dirty = YES;
    [self scheduleAutosave];
}

#pragma mark - NSTextViewDelegate

- (BOOL)textView:(NSTextView *)textView shouldChangeTextInRanges:(NSArray<NSValue *> *)affectedRanges replacementStrings:(nullable NSArray<NSString *> *)replacementStrings {
    return YES;
}

#pragma mark - Formatter Handler

- (IBAction)boldButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatBold];
}

- (IBAction)strikeThroughButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatStrikeThrough];
}

- (IBAction)italicButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatItalic];
}

- (IBAction)quoteButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatQuote];
}

- (IBAction)codeButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatCode];
}

- (IBAction)insertLinkButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatLink];
}

- (IBAction)listBulletedButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatListBulleted];
}

- (IBAction)listNumberedButtonClicked:(NSButton *)sender {
    [self applyFormat:TextConverterFormatListNumbered];
}

#pragma mark - Menu Handler

- (IBAction)newDocument:(id)sender {
    if (!_dirty) {
        [self newFile];
        return;
    }
    [self.dialogPresenter confirmDiscardingChangesForWindow:self.view.window completionHandler:^(NSModalResponse returnCode) {
        switch (returnCode) {
            case NSAlertFirstButtonReturn: {  // Save
                if ([self saveFile]) {
                    return;
                }
                [self promptForSaveURLWithCompletionHandler:^(BOOL result) {
                    if (result) {
                        [self newFile];
                    }
                }];
                break;
            }
            case NSAlertThirdButtonReturn:  // Don't Save
                [self newFile];
                break;
            case NSAlertSecondButtonReturn: // Cancel
            default:
                break;
        }
    }];
}

- (IBAction)openDocument:(id)sender {
    if (!_dirty) {
        [self promptForOpenURLWithCompletionHandler:nil];
        return;
    }
    [self.dialogPresenter confirmDiscardingChangesForWindow:self.view.window completionHandler:^(NSModalResponse returnCode) {
        switch (returnCode) {
            case NSAlertFirstButtonReturn: {  // Save
                if ([self saveFile]) {
                    return;
                }
                [self promptForSaveURLWithCompletionHandler:^(BOOL result) {
                    if (result) {
                        [self promptForOpenURLWithCompletionHandler:nil];
                    }
                }];
                break;
            }
            case NSAlertThirdButtonReturn:  // Don't Save
                [self promptForOpenURLWithCompletionHandler:nil];
                break;
            case NSAlertSecondButtonReturn: // Cancel
            default:
                break;
        }
    }];
}

- (IBAction)saveDocument:(id)sender {
    if ([self saveFile]) {
        return;
    }
    [self promptForSaveURLWithCompletionHandler:nil];
}

- (IBAction)saveDocumentAs:(id)sender {
    [self promptForSaveURLWithCompletionHandler:nil];
}

#pragma mark - Private Methods

- (NSArray<NSButton *> *)formattingButtons {
    NSArray<NSString *> *selectorNames = @[
        NSStringFromSelector(@selector(boldButtonClicked:)),
        NSStringFromSelector(@selector(strikeThroughButtonClicked:)),
        NSStringFromSelector(@selector(italicButtonClicked:)),
        NSStringFromSelector(@selector(quoteButtonClicked:)),
        NSStringFromSelector(@selector(codeButtonClicked:)),
        NSStringFromSelector(@selector(insertLinkButtonClicked:)),
        NSStringFromSelector(@selector(listBulletedButtonClicked:)),
        NSStringFromSelector(@selector(listNumberedButtonClicked:)),
    ];
    NSMutableArray<NSButton *> *buttons = [NSMutableArray array];
    for (NSString *selectorName in selectorNames) {
        NSButton *button = [self formattingButtonInView:self.view matchingSelectorName:selectorName];
        if (button) {
            [buttons addObject:button];
        }
    }
    return buttons;
}

- (NSButton *)formattingButtonInView:(NSView *)view matchingSelectorName:(NSString *)selectorName {
    for (NSView *subview in view.subviews) {
        if ([subview isKindOfClass:[NSButton class]]) {
            NSButton *button = (NSButton *)subview;
            if (NSStringFromSelector(button.action) && [NSStringFromSelector(button.action) isEqualToString:selectorName]) {
                return button;
            }
        }
        NSButton *button = [self formattingButtonInView:subview matchingSelectorName:selectorName];
        if (button) {
            return button;
        }
    }
    return nil;
}

- (void)layoutFormattingToolbar {
    NSRect bounds = self.view.bounds;
    NSScrollView *scrollView = self.textView.enclosingScrollView;
    if (!scrollView) {
        return;
    }

    NSArray<NSButton *> *buttons = [self formattingButtons];
    CGFloat top = NSMaxY(bounds) - EditorFormattingToolbarPadding;
    CGFloat buttonY = top - EditorFormattingButtonHeight;
    CGFloat x = EditorFormattingToolbarPadding;
    for (NSButton *button in buttons) {
        button.frame = NSMakeRect(x,
                                  buttonY,
                                  EditorFormattingButtonWidth,
                                  EditorFormattingButtonHeight);
        x += EditorFormattingButtonWidth + EditorFormattingButtonGap;
    }

    if (self.converterPopup) {
        CGFloat popupWidth = MIN(EditorConverterPopupWidth, MAX(0.0, NSWidth(bounds) - (2.0 * EditorFormattingToolbarPadding)));
        CGFloat popupX = MAX(x, NSWidth(bounds) - popupWidth - EditorFormattingToolbarPadding);
        self.converterPopup.frame = NSMakeRect(popupX,
                                               buttonY,
                                               popupWidth,
                                               EditorConverterPopupHeight);
    }

    CGFloat scrollViewHeight = MAX(0.0, buttonY - EditorFormattingToolbarPadding);
    scrollView.frame = NSMakeRect(0.0, 0.0, NSWidth(bounds), scrollViewHeight);
}

- (NSString *)loadSample {
    return ConverterManager.sharedInstance.selectedConverter.sample;
}

- (NSString *)documentBody {
    NSString *body = self.textView.string;
    if (!body) {
        body = @"Unknown";
    }
    return body;
}

- (void)replaceCharactersInRange:(NSRange)range withString:(NSString *)string {
    if ([self.textView shouldChangeTextInRange:range replacementString:string]) {
        [self.textView replaceCharactersInRange:range withString:string];
        [self.textView didChangeText];
        _dirty = YES;
    }
}

- (void)applyFormat:(TextConverterFormat)format {
    NSRange range = self.textView.selectedRange;
    if (range.length == 0) {
        return;
    }

    EditorTextFormatter *formatter = [[EditorTextFormatter alloc] initWithConverter:ConverterManager.sharedInstance.selectedConverter];
    NSRange selectedRange = range;
    NSString *updatedString = [formatter stringByApplyingFormat:format
                                                         toText:self.textView.string
                                                          range:range
                                                  selectedRange:&selectedRange];
    self.textView.string = updatedString;
    self.textView.selectedRange = selectedRange;
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
    _dirty = YES;
    [self scheduleAutosave];
}

- (NSString *)defaultDirectoryPath {
    NSString *path = _filePath;
    if (!path) {
        path = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, NO).firstObject;
    }
    return path;
}

- (void)startWatchingCurrentFile {
    __weak typeof(self) weakSelf = self;
    [self.fileWatcher watchFileAtPath:_filePath changeHandler:^{
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) {
            return;
        }
        [strongSelf reloadFileFromDiskIfNeeded];
    }];
}

- (void)scheduleAutosave {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(autosaveIfNeeded) object:nil];
    [self performSelector:@selector(autosaveIfNeeded) withObject:nil afterDelay:EditorAutosaveDelay];
}

- (void)autosaveIfNeeded {
    if (!_dirty || !_filePath) {
        return;
    }
    [self saveFile];
}

- (void)reloadFileFromDiskIfNeeded {
    if (_dirty || !_filePath) {
        return;
    }
    if (![self openFile]) {
        [self.fileWatcher stopWatching];
    }
}

- (void)promptForOpenURLWithCompletionHandler:(void (^ _Nullable)(BOOL result))handler {
    [self.dialogPresenter showOpenFilePanelForWindow:self.view.window
                                         initialPath:self.defaultDirectoryPath
                                   completionHandler:^(NSURL * _Nullable selectedURL) {
        BOOL success = NO;
        if (selectedURL) {
            self->_filePath = selectedURL.path;
            success = [self openFile];
        }
        if (handler) {
            handler(success);
        }
    }];
}

- (void)promptForSaveURLWithCompletionHandler:(void (^ _Nullable)(BOOL result))handler {
    [self.dialogPresenter showSaveFilePanelForWindow:self.view.window
                                         initialPath:self.defaultDirectoryPath
                                   completionHandler:^(NSURL * _Nullable selectedURL) {
        BOOL success = NO;
        if (selectedURL) {
            self->_filePath = selectedURL.path;
            success = [self saveFile];
        }
        if (handler) {
            handler(success);
        }
    }];
}

- (void)newFile {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(autosaveIfNeeded) object:nil];
    self.textView.string = @"New File";
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
    _filePath = nil;
    _dirty = NO;
    [self.fileWatcher stopWatching];
}

- (BOOL)openFile {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(autosaveIfNeeded) object:nil];
    if (!_filePath) {
        return NO;
    }
    BOOL isDirectory = NO;
    if (![NSFileManager.defaultManager fileExistsAtPath:_filePath isDirectory:&isDirectory]) {
        return NO;
    }
    if (isDirectory) {
        return NO;
    }
    EditorDocumentStore *store = [[EditorDocumentStore alloc] init];
    NSError *error = nil;
    NSString *string = [store readStringFromURL:[NSURL fileURLWithPath:_filePath] error:&error];
    if (!string) {
        return NO;
    }
    
    _dirty = NO;
    self.textView.string = string;
    [ConverterManager.sharedInstance setContentWithString:self.textView.string];
    [self startWatchingCurrentFile];
    return YES;
}

- (BOOL)saveFile {
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(autosaveIfNeeded) object:nil];
    if (!_filePath) {
        return NO;
    }
    if ([NSFileManager.defaultManager fileExistsAtPath:_filePath]) {
        NSError *error = nil;
        if (![NSFileManager.defaultManager removeItemAtPath:_filePath error:&error]) {
            return NO;
        }
    }
    
    _dirty = NO;
    EditorDocumentStore *store = [[EditorDocumentStore alloc] init];
    NSError *error = nil;
    BOOL success = [store writeString:self.textView.string toURL:[NSURL fileURLWithPath:_filePath] error:&error];
    if (success) {
        [self startWatchingCurrentFile];
    }
    return success;
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/EditorViewController.swift
import Cocoa

@objc(EditorViewController)
class EditorViewController: NSViewController, NSTextViewDelegate {
    private static let autosaveDelay: TimeInterval = 0.5
    private static let formattingButtonWidth: CGFloat = 32.0
    private static let formattingButtonHeight: CGFloat = 25.0
    private static let formattingButtonGap: CGFloat = 8.0
    private static let formattingToolbarPadding: CGFloat = 6.0
    private static let converterPopupWidth: CGFloat = 170.0
    private static let converterPopupHeight: CGFloat = 25.0

    @IBOutlet weak var textView: NSTextView!
    @IBOutlet weak var converterPopup: NSPopUpButton!

    @objc var filePath: String?
    @objc var dirty = false
    @objc lazy var dialogPresenter = EditorDialogPresenter()
    private lazy var fileWatcher = EditorFileWatcher()

    private var presentationWindow: NSWindow? {
        viewIfLoaded?.window
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        textView.setAccessibilityIdentifier("EditorTextView")
        converterPopup.setAccessibilityIdentifier("ConverterPopup")
        configureFormattingButtons()
        textView.string = loadSample()
        ConverterManager.shared.setContent(with: textView.string)
    }

    override func viewDidLayout() {
        super.viewDidLayout()
        configureFormattingButtons()
        layoutFormattingToolbar()
    }

    deinit {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(autosaveIfNeeded), object: nil)
        fileWatcher.stopWatching()
    }

    func textDidEndEditing(_ notification: Notification) {
        ConverterManager.shared.setContent(with: textView.string)
        dirty = true
        scheduleAutosave()
    }

    func textDidChange(_ notification: Notification) {
        ConverterManager.shared.setContent(with: textView.string)
        dirty = true
        scheduleAutosave()
    }

    func textView(
        _ textView: NSTextView,
        shouldChangeTextInRanges affectedRanges: [NSValue],
        replacementStrings: [String]?
    ) -> Bool {
        true
    }

    @IBAction func boldButtonClicked(_ sender: NSButton?) {
        applyFormat(.bold)
    }

    @IBAction func strikeThroughButtonClicked(_ sender: NSButton?) {
        applyFormat(.strikeThrough)
    }

    @IBAction func italicButtonClicked(_ sender: NSButton?) {
        applyFormat(.italic)
    }

    @IBAction func quoteButtonClicked(_ sender: NSButton?) {
        applyFormat(.quote)
    }

    @IBAction func codeButtonClicked(_ sender: NSButton?) {
        applyFormat(.code)
    }

    @IBAction func insertLinkButtonClicked(_ sender: NSButton?) {
        applyFormat(.link)
    }

    @IBAction func listBulletedButtonClicked(_ sender: NSButton?) {
        applyFormat(.listBulleted)
    }

    @IBAction func listNumberedButtonClicked(_ sender: NSButton?) {
        applyFormat(.listNumbered)
    }

    @IBAction func newDocument(_ sender: Any?) {
        if !dirty {
            newFile()
            return
        }
        dialogPresenter.confirmDiscardingChanges(for: presentationWindow) { [weak self] returnCode in
            guard let self else {
                return
            }
            switch returnCode {
            case .alertFirstButtonReturn:
                if self.saveFile() {
                    return
                }
                self.promptForSaveURL { result in
                    if result {
                        self.newFile()
                    }
                }
            case .alertThirdButtonReturn:
                self.newFile()
            default:
                break
            }
        }
    }

    @IBAction func openDocument(_ sender: Any?) {
        if !dirty {
            promptForOpenURL(completionHandler: nil)
            return
        }
        dialogPresenter.confirmDiscardingChanges(for: presentationWindow) { [weak self] returnCode in
            guard let self else {
                return
            }
            switch returnCode {
            case .alertFirstButtonReturn:
                if self.saveFile() {
                    return
                }
                self.promptForSaveURL { result in
                    if result {
                        self.promptForOpenURL(completionHandler: nil)
                    }
                }
            case .alertThirdButtonReturn:
                self.promptForOpenURL(completionHandler: nil)
            default:
                break
            }
        }
    }

    @IBAction func saveDocument(_ sender: Any?) {
        if saveFile() {
            return
        }
        promptForSaveURL(completionHandler: nil)
    }

    @IBAction func saveDocumentAs(_ sender: Any?) {
        promptForSaveURL(completionHandler: nil)
    }

    private var formattingButtons: [NSButton] {
        formattingButtonItems.compactMap { formattingButton(in: view, matchingSelectorName: NSStringFromSelector($0.selector)) }
    }

    private var formattingButtonItems: [(selector: Selector, toolTip: String)] {
        [
            (#selector(boldButtonClicked(_:)), "Bold"),
            (#selector(strikeThroughButtonClicked(_:)), "Strikethrough"),
            (#selector(italicButtonClicked(_:)), "Italic"),
            (#selector(quoteButtonClicked(_:)), "Block Quote"),
            (#selector(codeButtonClicked(_:)), "Code"),
            (#selector(insertLinkButtonClicked(_:)), "Insert Link"),
            (#selector(listBulletedButtonClicked(_:)), "Bulleted List"),
            (#selector(listNumberedButtonClicked(_:)), "Numbered List")
        ]
    }

    private func configureFormattingButtons() {
        let iconColor = toolbarIconColor()
        for item in formattingButtonItems {
            guard let button = formattingButton(in: view, matchingSelectorName: NSStringFromSelector(item.selector)) else {
                continue
            }
            if let image = button.image {
                button.image = tintedNonTemplateImage(from: image, color: iconColor)
            }
            button.imagePosition = .imageOnly
            button.contentTintColor = nil
            button.toolTip = item.toolTip
        }
    }

    private func toolbarIconColor() -> NSColor {
        let appearanceName = view.effectiveAppearance.bestMatch(from: [.darkAqua, .aqua])
        return appearanceName == .darkAqua
            ? NSColor(calibratedWhite: 0.86, alpha: 1.0)
            : NSColor(calibratedWhite: 0.20, alpha: 1.0)
    }

    private func tintedNonTemplateImage(from image: NSImage, color: NSColor) -> NSImage {
        let tintedImage = NSImage(size: image.size)
        tintedImage.lockFocus()
        image.draw(in: NSRect(origin: .zero, size: image.size), from: .zero, operation: .sourceOver, fraction: 1.0)
        color.setFill()
        NSRect(origin: .zero, size: image.size).fill(using: .sourceAtop)
        tintedImage.unlockFocus()
        tintedImage.isTemplate = false
        return tintedImage
    }

    private func formattingButton(in view: NSView, matchingSelectorName selectorName: String) -> NSButton? {
        for subview in view.subviews {
            if let button = subview as? NSButton,
               let action = button.action,
               NSStringFromSelector(action) == selectorName {
                return button
            }
            if let button = formattingButton(in: subview, matchingSelectorName: selectorName) {
                return button
            }
        }
        return nil
    }

    private func layoutFormattingToolbar() {
        let bounds = view.bounds
        guard let scrollView = textView.enclosingScrollView else {
            return
        }

        let top = bounds.maxY - Self.formattingToolbarPadding
        let buttonY = top - Self.formattingButtonHeight
        var x = Self.formattingToolbarPadding
        for button in formattingButtons {
            button.frame = NSRect(
                x: x,
                y: buttonY,
                width: Self.formattingButtonWidth,
                height: Self.formattingButtonHeight
            )
            x += Self.formattingButtonWidth + Self.formattingButtonGap
        }

        if let converterPopup {
            let popupWidth = min(Self.converterPopupWidth, max(0.0, bounds.width - (2.0 * Self.formattingToolbarPadding)))
            let popupX = max(x, bounds.width - popupWidth - Self.formattingToolbarPadding)
            converterPopup.frame = NSRect(
                x: popupX,
                y: buttonY,
                width: popupWidth,
                height: Self.converterPopupHeight
            )
        }

        let scrollViewHeight = max(0.0, buttonY - Self.formattingToolbarPadding)
        scrollView.frame = NSRect(x: 0.0, y: 0.0, width: bounds.width, height: scrollViewHeight)
    }

    private func loadSample() -> String {
        ConverterManager.shared.selectedConverter.sample ?? ""
    }

    @objc func documentBody() -> String {
        textView?.string ?? "Unknown"
    }

    @objc(replaceCharactersInRange:withString:)
    func replaceCharacters(in range: NSRange, with string: String) {
        guard textView.shouldChangeText(in: range, replacementString: string) else {
            return
        }
        textView.replaceCharacters(in: range, with: string)
        textView.didChangeText()
        dirty = true
    }

    @objc func applyFormat(_ format: TextConverterFormat) {
        let range = textView.selectedRange()
        guard range.length > 0 else {
            return
        }

        let formatter = EditorTextFormatter(converter: ConverterManager.shared.selectedConverter)
        var selectedRange = range
        let updatedString = formatter.stringByApplyingFormat(
            format,
            toText: textView.string,
            range: range,
            selectedRange: &selectedRange
        )
        textView.string = updatedString
        textView.setSelectedRange(selectedRange)
        ConverterManager.shared.setContent(with: textView.string)
        dirty = true
        scheduleAutosave()
    }

    @objc var defaultDirectoryPath: String {
        filePath ?? NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? NSHomeDirectory()
    }

    private func startWatchingCurrentFile() {
        fileWatcher.watchFile(atPath: filePath) { [weak self] in
            self?.reloadFileFromDiskIfNeeded()
        }
    }

    private func scheduleAutosave() {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(autosaveIfNeeded), object: nil)
        perform(#selector(autosaveIfNeeded), with: nil, afterDelay: Self.autosaveDelay)
    }

    @objc func autosaveIfNeeded() {
        guard dirty, filePath != nil else {
            return
        }
        _ = saveFile()
    }

    @objc func reloadFileFromDiskIfNeeded() {
        guard !dirty, filePath != nil else {
            return
        }
        if !openFile() {
            fileWatcher.stopWatching()
        }
    }

    @objc(promptForOpenURLWithCompletionHandler:)
    func promptForOpenURL(completionHandler handler: ((Bool) -> Void)?) {
        dialogPresenter.showOpenFilePanel(for: presentationWindow, initialPath: defaultDirectoryPath) { [weak self] selectedURL in
            guard let self else {
                handler?(false)
                return
            }
            var success = false
            if let selectedURL {
                self.filePath = selectedURL.path
                success = self.openFile()
            }
            handler?(success)
        }
    }

    @objc(promptForSaveURLWithCompletionHandler:)
    func promptForSaveURL(completionHandler handler: ((Bool) -> Void)?) {
        dialogPresenter.showSaveFilePanel(for: presentationWindow, initialPath: defaultDirectoryPath) { [weak self] selectedURL in
            guard let self else {
                handler?(false)
                return
            }
            var success = false
            if let selectedURL {
                self.filePath = selectedURL.path
                success = self.saveFile()
            }
            handler?(success)
        }
    }

    @objc func newFile() {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(autosaveIfNeeded), object: nil)
        textView.string = "New File"
        ConverterManager.shared.setContent(with: textView.string)
        filePath = nil
        dirty = false
        fileWatcher.stopWatching()
    }

    @objc func openFile() -> Bool {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(autosaveIfNeeded), object: nil)
        guard let filePath else {
            return false
        }
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: filePath, isDirectory: &isDirectory), !isDirectory.boolValue else {
            return false
        }

        let store = EditorDocumentStore()
        guard let string = try? store.readString(from: URL(fileURLWithPath: filePath)) else {
            return false
        }

        dirty = false
        textView.string = string
        ConverterManager.shared.setContent(with: textView.string)
        startWatchingCurrentFile()
        return true
    }

    @objc func saveFile() -> Bool {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(autosaveIfNeeded), object: nil)
        guard let filePath else {
            return false
        }
        if FileManager.default.fileExists(atPath: filePath) {
            do {
                try FileManager.default.removeItem(atPath: filePath)
            } catch {
                return false
            }
        }

        dirty = false
        let store = EditorDocumentStore()
        do {
            try store.writeString(textView.string, to: URL(fileURLWithPath: filePath))
            startWatchingCurrentFile()
            return true
        } catch {
            return false
        }
    }
}
